docs：add En docs (#30)
cryptape/overlord: docs/architecture_en.md

# Overlord

### Goal

> Overlord is a BFT consensus algorithm aiming to support thousands of transactions per second under hundreds of consensus nodes, with transaction delays of no more than a few seconds. Simply put, it is a high-performance consensus algorithm able to meets most of the real business needs.

Reviewer: ```suggestion
Overlord is a Byzantine fault tolerance (BFT) consensus algorithm aiming to support thousands of transactions per second under hundreds of consensus nodes, with transaction delays of no more than a few seconds. Simply put, it is a high-performance consensus algorithm able to meets most of the real business needs.
```

### Background

Usually, a consensus consists of at least two layers of semantics:

* Complete the transaction sequence
* Achieve consensus on the latest state

> For the blockchain of the UTXO model, the new state is implicit in the transaction output, so 1 and 2 are integral and inseparable. For the blockchain of the Account model, the transaction does not contain the state, and only after the transaction is executed can the latest state be generated and the state is saved in a independent MPT tree.

Reviewer: ```suggestion
For the blockchain of the UTXO model, the new state is implicit in the transaction output, so 1 and 2 are integral and inseparable. For the blockchain of the Account model, the transaction does not contain the state, and only after the transaction is executed can the latest state be generated and the state is saved in an independent MPT tree.
```

> For the Account model, in order to implement the second semantic, the common method is to let the consensus nodes execute all the transactions firstly. In this step, the latest state could be calculated and be saved to the blockheader. And then, consensus node will broadcast the block in which the transaction is sequenced.  Once the consensus is achieved,  the state and the transaction sequence is determined among all the nodes.  

Reviewer: ```suggestion
For the Account model, in order to implement the second semantic, the common method is to let the consensus nodes execute all the transactions firstly. In this step, the latest state could be calculated and be saved to the block header. And then, the consensus node will broadcast the block in which the transaction is sequenced.  Once the consensus is achieved,  the state and the transaction sequence is determined among all the nodes.  
```

> However, this method restricts the transaction processing capability of the BFT-like consensus algorithm. As shown in the figure below, Only after the block B(h)（ h means the block height） reaches a consensus, the new leader can pack and execute B（h+1），and then broadcast B（h+1）. After receiving B(h+1), other consensus nodes need to execute B(h+1) again to verify its correctness. So, in this process, these two serial block execution processes slowed down the consensus efficiency.

Reviewer: ```suggestion
However, this method restricts the transaction processing capability of the BFT-like consensus algorithm. As shown in the figure below, Only after the block B(h) (h means the block height) reaches a consensus, the new leader can pack and execute B(h+1)，and then broadcast B(h+1). After receiving B(h+1), other consensus nodes need to execute B(h+1) again to verify its correctness. So, in this process, these two serial block execution processes slowed down the consensus efficiency.
```

<div align=center><img src="./assets/block_process.png"></div>

An improved method is that the Leader does not execute the block immediately after packing the new block. After the block reaches a consensus, the consensus node executes the block to generate a new state, and the next height leader will pack this state and the next height blocks together to participate in the next consensus process. This method saves one block execution process time.

> When examining this improvement from a more microscopic perspective, we found that there is still much room for improvement. This is because the consensus module and the execution module of any consensus node are always serial throughout the consensus process. As shown in the figure above, when the consensus module runs, the execution module is always idle, and vice versa. . If the execution module and the consensus module can be paralleled, the consensus transaction processing capability can theoretically reach the maximum processing limit of the execution module.

Reviewer: ```suggestion
When examining this improvement from a more macroscopic perspective, we found that there is still much room for improvement. This is because the consensus module and the execution module of any consensus node are always serial throughout the consensus process. As shown in the figure above, when the consensus module runs, the execution module is always idle, and vice versa. If the execution module and the consensus module can be paralleled, the consensus transaction processing capability can theoretically reach the maximum processing limit of the execution module.
```

### Protocol description

In Overlord, a consensus process is called an epoch. The epoch contains two parts, Header and Body (as shown below). The core structure of epoch is shown below, `epoch_id` is a monotonically increasing value, equivalent to height; `prev_hash` is the hash of the previous epoch; `order_root` is the merkle root of all pending transactions contained in the Body; `state_root` represents the latest world The MPT root of the state; `confirm_roots` represents the `order_root` collection from the `state_root` of the previous epoch to the `state_root` of the current epoch; the `receipt_roots` records the `receipt_root` corresponding to each `order_root` being executed; proof is proof of the previous epoch .

<div align=center><img src="./assets/epoch.png"></div>

> In this method, the consensus module batches the transactions to make a consensus. After the consensus is reached, the ordered transaction set is added to the queue to be executed, and the execution module executes in order of the transaction set, and each execution of the transaction set is performed. The order_root of the transaction set to be executed, and the executed stateRoot are sent to the consensus module. When packing the transactions to assemble the epoch, the leader take the latest state_root as the latest state to participate in the consensus.

Reviewer: ```suggestion
In this method, the consensus module batches the transactions to make a consensus. After the consensus is reached, the ordered transaction set is added to the queue to be executed, and the execution module executes in order of the transaction set, and each execution of the transaction set is performed. The ordered root of the transaction set to be executed, and the executed state roots are sent to the consensus module. When packing the transactions to assemble the epoch, the leader takes the latest state root as the latest state to participate in the consensus.
```

Overlord is an interpretation layer above the specific consensus algorithm. By reinterpreting the semantics of consensus, the transaction sequence is decoupled from the state consensus, so that higher transaction processing capability can be obtained in actual operation. In theory, Overlord can be based on almost any BFT-like consensus algorithm, specifically in our project based on the improved Tendermint.

We have made three major improvements compared to Tendermint:

> 1. Apply the aggregate signature to Tendermint to make the consensus message complexity from <img src="https://latex.codecogs.com/svg.latex?\inline&space;O(n^{2})" title= "O(n^{2})" /> falls to <img src="https://latex.codecogs.com/svg.latex?\inline&space;O(n)" title="O(n)" / >, thus being able to support more consensus nodes

Reviewer: ```suggestion
1. Apply the aggregate signature to Tendermint to make the consensus message complexity from <img src="https://latex.codecogs.com/svg.latex?\inline&space;O(n^{2})" title= "O(n^{2})" /> falls to <img src="https://latex.codecogs.com/svg.latex?\inline&space;O(n)" title="O(n)" />, thus being able to support more consensus nodes
```

#### Verify parallelism

After receiving the *proposal*, the consensus node will verify the *CompactEpoch* (to obtain the complete transaction and verify the correctness of the transaction) in parallel with the *prevote* vote. Only after receiving the *prevote* aggregate signature and the *CompactEpoch* test result will the *precommit* be cast.

## Overlord architecture

The Overlord consensus consists of the following components:

* State Machine (SMR): State transition based on input messages
* State storage (State): used to store the status of the proposal, voting, etc.
* Timer: Set the timeout period to trigger the state machine operation
* Wal: used to read and write Wal logs

> In the Overlord consensus architecture, when a message is received, the state storage module performs a basic check on the message. After passing, the status is updated according to the received message and the message is transmitted to the state machine. In addition, a timer is required to maintain activity, and when the timer expires, the timer calls the interface to trigger the state machine. The state machine will throw a current state event after making the state change. The state storage module and the timer module listen to the event thrown by the state machine, and perform corresponding processing according to the monitored event, such as writing Wal, sending a vote, setting timing. And so on. At the time of the restart, the state storage module reads the data from the Wal and sends it to the state machine. The overall architecture is shown below:

Reviewer: ```suggestion
In the Overlord consensus architecture, when a message is received, the state storage module performs a basic check on the message. After passing, the status is updated according to the received message and the message is transmitted to the state machine. In addition, a timer is required to maintain activity, and when the timer expires, the timer calls the interface to trigger the state machine. The state machine will throw a current state event after making the state change. The state storage module and the timer module listen to the event thrown by the state machine, and perform corresponding processing according to the monitored event, such as writing Wal, sending a vote, setting the timing. And so on. At the time of the restart, the state storage module reads the data from the Wal and sends it to the state machine. The overall architecture is shown below:
```

<div align=center><img src="./assets/arch_overlord.png"></div>

### Consensus State Machine (SMR)

The state machine module is the logical core of the entire consensus, and its main functions are state changes and *lock* control. When the received message is triggered, the status change is made according to the received message, and the changed status is thrown as an event. In our implementation, Overlord uses a Tendermint state machine that applies BLS aggregate signature optimization for consensus. The overall working process is as follows.

> #### Prepare phase

Reviewer: ```suggestion
#### Propose step
```

The node uses a deterministic random algorithm to determine the leader of the current round.

*Leader*: Broadcast a proposal

*Others*: Set a timer T1 to send a *prevote* vote to *Relayer* when the proposal is received

> #### Pre-voting phase

Reviewer: ```suggestion
#### Prevote step
```

*Relayer*: Set a timer T2 to aggregate the received *prevote* votes and generate a bitmap to broadcast the aggregated votes and bitmaps to other nodes.

*Others*: Set a timer T2, check the validity of the aggregated prevote vote, generate *PoLC* send precommit vote

> ####Verification phase

Reviewer: ```suggestion
####Verification step
```

All nodes set a timer T3. After receiving the verification result of the *proposal*, they enter the *pre-commit* stage.

> #### Pre-commit phase

Reviewer: ```suggestion
#### Precommit step
```

*Relayer*: Set a timer T4 to aggregate the received precommit votes and generate a bitmap to broadcast the aggregated votes and bitmaps to other nodes.

*Others*: Set a timer T4 to check the legitimacy of the aggregated precommit vote

> #### Commit stage

Reviewer: ```suggestion
#### Commit step
```

Author: 我看 hotstuff 的 paper 里面用的 phase 这个词：https://www.cs.unc.edu/~reiter/papers/2019/PODC.pdf

Reviewer: The overlord is not based on hotstuff.

```
pub struct SMRTrigger {
    pub hash: Hash,
    pub round: Option<u64>,
    pub trigger_type: TriggerType,
}
```

The output structure of the state machine is as follows:

```
pub enum SMREvent {
    /// New round event
    /// for state: update round,
    /// for timer: set a propose step timer.
    NewRoundInfo {
        round: u64,
        lock_round: Option<u64>,
        lock_proposal: Option<Hash>,
    },
    /// Prevote event,
    /// for state: transmit a prevote vote,
    /// for timer: set a prevote step timer.
    PrevoteVote(Hash),
    /// Precommit event,
    /// for state: transmit a precommit vote,
    /// for timer: set a precommit step timer.
    PrecommitVote(Hash),
    /// Commit event
    /// for state: do commit,
    /// for timer: do nothing.
    Commit(Hash),
    /// Stop event,
    /// for state: stop process,
    /// for timer: stop process.
    Stop,
}
```

#### State Machine Interface

```
/// Create a new SMR service.
pub fn new() -> Self
/// Trigger a SMR action.
pub fn trigger(&self, gate: SMRTrigger) -> Result<(), Error>
/// Goto a new consensus epoch.
pub fn new_epoch(&self, epoch_id: u64) -> Result<(), Error>
```

### State storage (State)

> The state storage module is the functional core of the entire consensus. The main functions are storage state, message distribution, block and cryptography related operations. In the working process, for the message transmitted by the network layer, the first check is performed to verify the validity of the message. Determine whether the written message needs to be written to Wal. The message is then sent to the state machine. The state storage module constantly listens for events thrown by the state machine and processes them accordingly.

Reviewer: ```suggestion
The state storage module is the functional core of the entire consensus. The main functions are storage state, message distribution, block, and cryptography related operations. In the working process, for the message transmitted by the network layer, the first check is performed to verify the validity of the message. Determine whether the written message needs to be written to Wal. The message is then sent to the state machine. The state storage module constantly listens for events thrown by the state machine and processes them accordingly.
```
